rfcs/text/0008-threat-intel.md

```diff
@@ -23,30 +23,38 @@ Stage 1: Describe at a high level how this change affects fields. Which fieldset
 
 ### Proposed New Fields for Threat fieldset
 
-  * threat.time_first_seen
+  * threat.ioc.time_first_seen
     _The date and time when intelligence souce first reported sighting this indicator._
-  * threat.time_last_seen
+  * threat.ioc.time_last_seen
     _The date and time when intelligence source last reported sighting this indicator._
-  * threat.sightings
+  * threat.ioc.sightings
     _Number of times this indicator was observed conducting threat activity._
   * threat.type
     _Type of indicator as reprsented by Cyber Observable in STIX 2.0_
-  * threat.description
+  * threat.ioc.description
     _Describes the type of action conducted by the threat._
-  * threat.tlp
+  * threat.ioc.tlp
     _Traffic Light Protocol, which dictates sharing policies_
-  * threat.classification
+  * threat.ioc.classification
     _Describes type of threat delivery (Hacktool etc.) and family name.
-  * threat.scanner_stats
+  * threat.ioc.scanner_stats
     _Count of Anti virus/EDR that successfully detected malicious file or URL. Sources like VirusTotal, Reversing Labs often provide these statistics._
 
-### Proposed New Values
+### Proposed New Values for Event Fieldset
 
   * event.kind:enrichment _Propose adding this value to capture outcome of this event. It could also appy to other types of contextual data such as directory services, IPAM data, asset lists._
   * event.category:threat _Proposed threat.type field would be a subcategory for this value of event.category_
   * event.type:indicator _Proposed value represents type of threat information. In future this could be extended to other STIX 2.0 Standard Data Objects like Actor, Infrastucture etc._
 
-### Existing ECS Fields For Nested Use in Threat.*
+### Using existing Event Fieldset
+ * event.reference _URL to the intelligence source_
+ * event.provider _name of intelligence provider_
+ * event.dataset _name of specific dataset from the intelligence source. Intelligence sources often provide multiple datasets - IP blocklist, File hash blocklist etc.
+ * event.severity _severity provided by threat intelligence source_
+ * event.risk_score _risk score provided by threat intelligence source_
+ * event.original _raw intelligence event_
+
+### Using existing ECS Fields nested under Threat.ioc.*
```

Comment:
How about also adding the `process` fields to the possibilities of being nested?

For example a given `process.args` could be an IOC

Reply:
I originally dropped this in slack, but I think it got lost. 

on a bigger picture of "intel words and semantics", it makes sense to me that something like `W32.Trojan.22gp.1201` as labelled perhaps from an AV engine or malware analysis feed (ala VirusTotal, ReversingLabs, etc) should be assigned as a `rule.name`. The rationale being that there's clearly some sort of signature, whether atomic or behavioral.

Meanwhile, saying something is a variant of a given malware family is informative and takes raw information and transforms it into intelligence and better aligns with ontologies like STIX or MAEC. I'm proposing something like:

```
{
"file": { "hash": "abcdef01234567890...", "name": "totally_bad.exe", ...},
"rule": [ {"author": "ALYac", "name": "Generic.Malware.SPV!PKprn.5A432451", "update": "20200915"} ],
"threat": { 
     "malware": {"name": "WickedBadness.A", "family": "WickedBadness"}, 
     "intrusion_set": {"name": "APT127", ...},
}
```

Now, that may not align where things are currently heading and I'm not trying to open pandoras box here... the above approach does a few things, IMO:

1. This can allow me to enrich data coming from various sources (i.e. my VT feed, or file access data from an endpoint, or file transfer data from my proxy or IDS. The enrichment takes previous knowledge and tags this file, network, or process event with "threat" knowledge.
2. Allows me to take a rule (IDS, AV, Elastic Signal, etc) and say that "X" indicates `threat.malware.name: "WickedBadness.A"`, taking normal time series events and making the detection actionable with why we care about said detection
3. If the document only included the subset of information to provides a `file.hash.sha256` and `threat.malware.name`, the hash is immediately actionable in how we apply filters in Elasticsearch becomes a pivot field that is shared with my logging sources directly. I can click a single visualization on a Kibana dashboard and drill into documents about a given file transfer and see in an adjacent visualization that the hash I clicked on is associated with a malware family or intrusion set.
Alternatively, nesting the known file threat indicators under `threat.file.hash`, we either use lookup detection rules (which we could do in either case), or we rely on a 3rd pivot field of `related.hash`, which duplicates data.






just sharing some thinking...  would love to discuss more in depth when the time is right. Right now may be premature, and if so...  sorry for stirring the pot

Reply:
Just to distill this a bit, the IOCs should to be actionable entities - an indicator. So `threat.malware.*` is context and descriptors of the IOC, but it's not an IOC in itself.

So while a lot of the fields are proposed under `threat.ioc.`, `threat.malware.{name,family,type}` makes sense to me to describe the indicator, which would likely be a hash value.

Reply:
https://github.com/elastic/ecs/pull/1127/files#r527125145

Reply:
Discussed this in depth with @shimonmodi and got some additional context. I think the `threat.ioc.*` makes sense from a enrichment perspective, but I think as stored in the "threat intel" index, keeping the shared top-level ECS fields makes the data more actionable from a retro-hunt perspective. Namely, if we get intelligence about events after they pass through the enrichment pipeline, those events will miss the threat data. Keeping them in the non-ioc fields allows hunting using dashboards and discover. Copying the respective enrichment match to `threat.ioc.*` keep the original document intact while also giving context of why we think this particular event is a threat.

I want to propose some hypothetical documents and how this might work. Visualizing what the data could be helps me better understand what we're trying to achieve. Here's an example process start event with file metadata (I added the file hash for the sake of for example):

```json
{
  "agent": {
    "id": "0829aba6-34db-de36-1d42-30eac745e980",
    "type": "endpoint",
    "version": "7.10.0"
  },
  "process": {
    "name": "svchost.exe",
    "pid": 1644,
    "entity_id": "MDgyOWFiYTYtMzRkYi1kZTM2LTFkNDItMzBlYWM3NDVlOTgwLTE2NDQtMTMyNDk3MTA2OTcuNDc1OTExNTAw",
    "executable": "C:\\Windows\\System32\\svchost.exe"
  },
  "message": "Endpoint file event",
  "@timestamp": "2020-11-17T19:07:46.0956672Z",
  "file": {
    "path": "C:\\Windows\\Prefetch\\SVCHOST.EXE-AE7DB802.pf",
    "extension": "pf",
    "name": "SVCHOST.EXE-AE7DB802.pf",
    "hash": {
      "sha1": "bfb7759a67daeb65410490b4d98bb9da7d1ea2ce"
    }
  },
  "ecs": {
    "version": "1.5.0"
  },
  "data_stream": {
    "namespace": "default",
    "type": "logs",
    "dataset": "endpoint.events.file"
  },
  "host": {
    "hostname": "WinDev2001Eval",
    "os": {
      "Ext": {
        "variant": "Windows 10 Enterprise Evaluation"
      },
      "kernel": "1909 (10.0.18363.1139)",
      "name": "Windows",
      "family": "windows",
      "version": "1909 (10.0.18363.1139)",
      "platform": "windows",
      "full": "Windows 10 Enterprise Evaluation 1909 (10.0.18363.1139)"
    },
    "ip": [
      "192.168.93.145",
      "10.203.18.111",
      "fe80::4402:db6b:fd75:544",
      "127.0.0.1",
      "::1"
    ],
    "name": "WinDev2001Eval",
    "id": "5baae4dd-4abe-4ba5-a0fb-704d6e7a4328",
    "mac": [
      "00:0c:29:76:2b:0a",
      "00:ff:7b:c6:4a:64"
    ],
    "architecture": "x86_64"
  },
  "event": {
    "sequence": 186465,
    "ingested": "2020-11-17T19:08:01.275480865Z",
    "created": "2020-11-17T19:07:46.0956672Z",
    "kind": "event",
    "module": "endpoint",
    "action": "creation",
    "id": "LurtOw/d18obyz+u++++/boE",
    "category": [
      "file"
    ],
    "type": [
      "creation"
    ],
    "dataset": "endpoint.events.file"
  },
  "user": {
    "domain": "NT AUTHORITY",
    "name": "SYSTEM",
    "id": "S-1-5-18"
  }
}
```
Meanwhile you have the following document containing threat intel:
```json
{
  "file": {
    "hash": {
      "sha1": "bfb7759a67daeb65410490b4d98bb9da7d1ea2ce"
    }
  },
  "threat": {
    "malware": {
      "name": "CryptoMinerX.B",
      "family": "CryptoMinerX",
      "type": "cryptominer"
    },
    "threat_actor": {
      "name": "CryptoCurrency R Us",
      "type": [
        "criminal"
      ]
    }
  }
}
```

What I'd like to see then, is an enriched document like so (abbreviated):

```json
{
  "process": {
    "name": "svchost.exe",
    "pid": 1644,
    "entity_id": "MDgyOWFiYTYtMzRkYi1kZTM2LTFkNDItMzBlYWM3NDVlOTgwLTE2NDQtMTMyNDk3MTA2OTcuNDc1OTExNTAw",
    "executable": "C:\\Windows\\System32\\svchost.exe"
  },
  "message": "Endpoint file event",
  "@timestamp": "2020-11-17T19:07:46.0956672Z",
  "file": {
    "path": "C:\\Windows\\Prefetch\\SVCHOST.EXE-AE7DB802.pf",
    "extension": "pf",
    "name": "SVCHOST.EXE-AE7DB802.pf",
    "hash": {
      "sha1": "bfb7759a67daeb65410490b4d98bb9da7d1ea2ce"
    }
  },
  ...
  "threat": {
    "ioc": {
      "file": {
        "hash": {
          "sha1": "bfb7759a67daeb65410490b4d98bb9da7d1ea2ce"
        }
      }
    },
    "malware": {
      "name": "CryptoMinerX.B",
      "family": "CryptoMinerX",
      "type": "cryptominer"
    },
    "threat_actor": {
      "name": "CryptoCurrency R Us",
      "type": [
        "criminal"
      ]
    }
  }
}
```
---
**DISCUSSION**

Keeping the indicator in what I'll call the "actionable" ECS field (e.g. `file.hash.sha1`), allows me to search across all documents, including my logs and threat data. I can apply a single filter on that hash field. I can then visualize them all in a common dashboard to find events that occurred prior to receiving new threat intel.

Copying the matched IOC to the `threat.ioc` object in the enriched document answers the analyst question of "what specific information from this event indicates that it is a cryptominer?" It could have just as easily been the process arguments (missing from this example) or the path.

This approach may complicate using the `enrich` processor on ingest. Namely, if we want to enrich the document with everything under `threat.*`, it's not just a single lookup. You'd have to lookup, then rename some fields, which isn't awful, I think. Alternatively, the threat feed indexes could make `file.hash.sha1` as an alias to `threat.ioc.file.hash.sha1`. This would require an alias for each indicator type field.

Digging into the specifics of using the enrich processor, how would it handle multiple IOC matches? Say if we had an IOC that was `{"user": {"id": "S-1-5-18"}}`, which is obviously concoted, but ideally, I'd like to see the merging of threat data into one. It might imply a criminal crypto mining threat actor, and also a nation-state threat actor that's leveraging the tool with some other attack pattern.

Reply:
Reading through this I see there are two "use-cases" for the threat fields which should both be supported.

1. Threat intelligence feeds stored in a dedicated index to be used by:
    * Enrichment processors to enrich documents when they enter Elasticsearch
    * The [indicator match detection](https://www.elastic.co/guide/en/security/current/rules-ui-create.html#create-indicator-rule) in the Security Solution
2. Enrich documents with threat information (enrichment processor) or logstash.

_Note: there is overlap in the two, this is dependend on where the enrichment information is stored/retrieved from_

Going trough some of the comments made:
**rule.name field**
>on a bigger picture of "intel words and semantics", it makes sense to me that something like W32.Trojan.22gp.1201 as labelled perhaps from an AV engine or malware analysis feed (ala VirusTotal, ReversingLabs, etc) should be assigned as a rule.name. The rationale being that there's clearly some sort of signature, whether atomic or behavioral.

For both use-cases I believe this information to be beneficial, however not in the `rule.name` field, as this could cause confusion with the existing detection engine `signal.rule.name` field. Perhaps a field like `threat.descriptor` which is an array of strings? this could then house not just the name of malware, but other descripting strings.  
This applies to the entire proposed `rule.*` structure, which is just a matter of (re)naming the field to prevent conflict/confusion.

**threat.malware field**
_duplicate response but for completion sake_  
The `threat.ioc` fields allow for a variety of threat indicators to be housed, indeed most of them can be linked to a type of malware, but not all. To this end I feel the `threat.malware` fields limits usage in comparison. Perhaps if the fields proposed under `malware` or the field itself is housed under `threat.ioc` it can be added without limiting usage?

What I am trying to convey is whereas (almost) all malware can be combined with an ioc not all ioc's can be combined with malware, e.g: ip addresses.

(_and ofcourse the threat.ioc field is my own proposal so i'm rooting for it_ 😉 )

**field usage**
>Discussed this in depth with @shimonmodi and got some additional context. I think the threat.ioc.* makes sense from a enrichment perspective, but I think as stored in the "threat intel" index, keeping the shared top-level ECS fields makes the data more actionable from a retro-hunt perspective. Namely, if we get intelligence about events after they pass through the enrichment pipeline, those events will miss the threat data. Keeping them in the non-ioc fields allows hunting using dashboards and discover. Copying the respective enrichment match to threat.ioc.* keep the original document intact while also giving context of why we think this particular event is a threat.

Well put, this will enable bot use-cases as well as allowing security practitioners to choose if they want to use the top level ECS fields in their queries or the ioc fields. 

**threat.threat_actor field**
The current usage of the `threat` fields is to map events to MITRE ATT&CK; Adding threat actor information can be beneficial to an analyst for faster on track analyses and research. Especially enabling them to look for TTP's. 
I do believe there should two requirements for this field:  
1. It should be named and mapped to [MITRE groups](https://attack.mitre.org/groups/) to stay in line with the MITRE allignment.
2. It should be able to house multiple actors (groups).

### Final note
I am really struggling to put my thoughts into words and not all sentences might be worded as properly as appropriate.

Reply:
Thanks everyone for the thought going into this 🤔 👍 

I like the symmetry of having the various fields that could be used to describe an IOC be directly the ECS fields. However I wonder if this would work in practice.

If an IOC is a "bad IP" e.g. the essential information is the IP. But in the IOC index, should the value be captured in `source.ip`, `destination.ip`, `client.ip` or `server.ip`? All we need to do is capture "the IP" itself, then we look for it in the appropriate places, depending on the event source.

We may have a similar problem with TLS or x509 certs, where in ECS we currently have fields to describe TLS exchanges, potentially mutual TLS, so `tls.client.*` and `tls.server.*`, but if e.g. a cipher or hash is bad no matter on which side it is, which field should be used to capture it in the IOC descriptor?

To get back to the symmetry of being able to filter for an IOC's indicator and get both the IOC descriptor and actual events show up, I'm not actually sure this saves users anything. In Elasticsearch, unlike SQL, we have to be explicit in which values we want to match. We can't say like in SQL "give me all events and indicators that have the same value in `source.ip`" for example. If you have 10 fresh new bad IPs to test for, you'd have to explicitly make an include query for `["bad ip 1", "bad ip 2", ...]`.

The bullet list below (`file.*`, `file.hash.*`, `url.*`...) is useful to give us a general idea of the landscape of the types of IOCs. I think a more concrete list of the actual descriptors / fields we expect to need would be helpful in figuring out the way forward.

Reply:
> During enrichment you can then "add" the threat.ioc.* fields to tell the story of what was the match.

I think this makes a lot of sense; knowing if the match was an "IP" or a "URL" or a "hash" is important.

Reply:
@dcode @peasead  -   based on today's discussion with @MikePaquette our next step is to move all actionable IOC information in threat intel documents to the top level, and not nest them under `threat.ioc.*` (as is currently proposed). we will be using `threat.ioc.*` for `file`, `url` etc. for enrichment use case.

Reply:
> The bullet list below (`file.*`, `file.hash.*`, `url.*`...) is useful to give us a general idea of the landscape of the types of IOCs. I think a more concrete list of the actual descriptors / fields we expect to need would be helpful in figuring out the way forward.

`threat.yml` and the table should now have descriptors and fields requested.
https://github.com/elastic/ecs/blob/1347f2ba4c0ef6c00d5ffccee7fa2ad854f631d0/rfcs/text/0008/threat.yml
https://github.com/elastic/ecs/blob/1347f2ba4c0ef6c00d5ffccee7fa2ad854f631d0/rfcs/text/0008-threat-intel.md

Reply:
@ebeahan is going to look at how non-direction network artifacts should be reflected, taking ECS as-a-whole into account in addition to `threat.*` and the enriched document.